File: client/go/loggingclient/BaseLogger.go
package loggingclient

import (
	"container/list"
	"fmt"
	"time"
)

// BaseLogger ILog 实现基类
type BaseLogger struct {
	Source string
}

var _block *TimerActionBlock

func getBlock() *TimerActionBlock {
	if _block == nil {
		_block = NewTimerActionBlock(
			send,
			LoggingSettings.QueueMaxLength,
			LoggingSettings.BufferSize,
			LoggingSettings.BlockElapsed)
		go _block.dequeue()
		fmt.Println(time.Now().String() + "初始化日志队列")
	}
	return _block
}

func send(buffer list.List) {
	logSender := getLogSender()
	logSender.send(buffer)
	loom := NewLogOnOffManager()
	loom.RefreshLogOnOff()

}

// NewBaseLogger 实例化一个BaseLogger引用对象
func NewBaseLogger(source string) *BaseLogger {
	var baseLogger = new(BaseLogger)
	baseLogger.Source = source
	//baseLogger.LoggingEnabled = loggingEnabled()
	return baseLogger
}

// Debug 实现ILog.Debug
func (bl *BaseLogger) Debug(params ...interface{}) {
	title, message, tags := bl.getLogParams(params)
	bl.log(title, message, tags, 1)
}

// Info 实现ILog.Info
func (bl *BaseLogger) Info(params ...interface{}) {
	title, message, tags := bl.getLogParams(params)
	bl.log(title, message, tags, 2)
}

// Warm 实现ILog.Warm
func (bl *BaseLogger) Warm(params ...interface{}) {
	title, message, tags := bl.getLogParams(params)
	bl.log(title, message, tags, 3)
}

// Error 实现ILog.Error
func (bl *BaseLogger) Error(params ...interface{}) {
	title, message, tags := bl.getLogParams(params)
	bl.log(title, message, tags, 4)
}

// Debugf 实现ILog.Debugf
func (bl *BaseLogger) Debugf(format string, params ...interface{}) {
	message := fmt.Sprintf(format, params...)
	bl.log("", message, nil, 1)
}

// Infof 实现ILog.Infof
func (bl *BaseLogger) Infof(format string, params ...interface{}) {
	message := fmt.Sprintf(format, params...)
	bl.log("", message, nil, 2)
}

// Warmf 实现ILog.Warmf
func (bl *BaseLogger) Warmf(format string, params ...interface{}) {
	message := fmt.Sprintf(format, params...)
	bl.log("", message, nil, 3)
}

// Errorf 实现ILog.Errorf
func (bl *BaseLogger) Errorf(format string, params ...interface{}) {
	message := fmt.Sprintf(format, params...)
	bl.log("", message, nil, 4)
}

// Metric 实现ILog.Metric
func (bl *BaseLogger) Metric(name string, value float64, tags map[string]string) {
	if !LoggingSettings.LoggingEnabled {
		return
	}
	var Metric = NewMetricEntity()
	Metric.Name = name
	Metric.Value = value
	Metric.Tags = tags
	Metric.Time = time.Now().Unix()
	var block = getBlock()
	block.enqueue(Metric)
}

// Metric2 实现ILog.Metric2
func (bl *BaseLogger) Metric2(name string, value float64, tags map[string]string) {
	if !LoggingSettings.LoggingEnabled {
		return
	}
	var Metric = NewMetricEntity()
	Metric.Name = name
	Metric.Value = value
	Metric.Tags = tags
	Metric.Time = time.Now().Unix()
	var block = getBlock()
	block.enqueue(Metric)
}

func (bl *BaseLogger) getLogParams(params interface{}) (title string, message string, tags map[string]string) {
	var _title = ""
	var _message = ""
	var _tags map[string]string
	var _parmas = params.([]interface{})
	var argLen = len(_parmas)
	if argLen == 1 {
		_message = bl.getLogParamStr(_parmas[0])
	} else if argLen == 2 {
		_title = bl.getLogParamStr(_parmas[0])
		_message = bl.getLogParamStr(_parmas[1])
	} else if argLen == 3 {
		_title = bl.getLogParamStr(_parmas[0])
		_message = bl.getLogParamStr(_parmas[1])
		_tags = _parmas[2].(map[string]string)
	} else {
		_title = bl.getLogParamStr(_parmas[0])
		_message = fmt.Sprint(_parmas)
	}
	return _title, _message, _tags
}

func (bl *BaseLogger) getLogParamStr(param interface{}) (str string) {
	switch param.(type) {
	case string:
		str = param.(string)
	case error:
		str = param.(error).Error()
	default:
		str = fmt.Sprint(param)
	}
	return str
}

func (bl *BaseLogger) createLog(source string, title string, message string, tags map[string]string, level int8) (ret *LogEntity) {
	var log = NewLogEntity()
	log.Level = level
	log.Message = message
	log.Tags = tags
	log.Time = time.Now().UnixNano() / 100
	log.Title = title
	log.Source = source
	return log
}

func (bl *BaseLogger) log(title string, message string, tags map[string]string, level int8) {
	if !LoggingSettings.LoggingEnabled {
		return
	}
	var loom = NewLogOnOffManager()
	var onOff = loom.getLogOnOff()
	if level == 1 && onOff.Debug != 1 {
		return
	}
	if level == 2 && onOff.Info != 1 {
		return
	}
	if level == 3 && onOff.Warm != 1 {
		return
	}
	if level == 4 && onOff.Error != 1 {
		return
	}
	var log = bl.createLog(bl.Source, title, message, tags, level)
	var block = getBlock()
	block.enqueue(log)
}

File: client/csharp/Logging.Client/Logger/BaseLogger.cs
using Logging.Client.LogSender;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Diagnostics;
using System.Net;
using System.Text;
using System.Threading;
using System.Web;

namespace Logging.Client
{
    internal abstract partial class BaseLogger : ILog
    {
        private static ITimerActionBlock<ILogEntity> block;

        private string Source { get; set; }

        public BaseLogger(string source = "")
        {
            this.Source = source;
        }

        /// <summary>
        /// 客户端是否启用日志
        /// </summary>
        private readonly static bool LoggingEnabled = Convert.ToBoolean(ConfigurationManager.AppSettings["LoggingEnabled"] ?? Settings.LoggingEnabled.ToString());


        private static LogSenderBase sender = LogSenderManager.GetLogSender();

        static BaseLogger()
        {
            if (!LoggingEnabled) { return; }

            int LoggingTaskNum = Convert.ToInt32(ConfigurationManager.AppSettings["LoggingTaskNum"] ?? Settings.LoggingTaskNum.ToString());

            int LoggingQueueLength = Convert.ToInt32(ConfigurationManager.AppSettings["LoggingQueueLength"] ?? Settings.LoggingQueueLength.ToString());

            int LoggingBufferSize = Convert.ToInt32(ConfigurationManager.AppSettings["LoggingBufferSize"] ?? Settings.LoggingBufferSize.ToString());

            int LoggingBlockElapsed = Convert.ToInt32(ConfigurationManager.AppSettings["LoggingBlockElapsed"] ?? Settings.LoggingBlockElapsed.ToString());

            if (LoggingTaskNum <= 0) { LoggingTaskNum = Settings.DefaultLoggingTaskNum; }

            if (LoggingQueueLength <= 0) { LoggingQueueLength = Settings.DefaultLoggingQueueLength; }

            if (LoggingBufferSize <= 0) { LoggingBufferSize = Settings.DefaultLoggingBufferSize; }

            if (LoggingBlockElapsed <= 0) { LoggingBlockElapsed = Settings.DefaultLoggingBlockElapsed; }



            if (LoggingTaskNum == 1)
            {
                block = new TimerActionBlock<ILogEntity>((buffer) =>
                {
                    Send(buffer);
                }, LoggingQueueLength, LoggingBufferSize, LoggingBlockElapsed);
            }
            else
            {
                block = new ThreadedTimerActionBlock<ILogEntity>(LoggingTaskNum, (buffer) =>
                {
                    Send(buffer);
                }, LoggingQueueLength, LoggingBufferSize, LoggingBlockElapsed);
            }
        }

        private static void Send(IList<ILogEntity> buffer)
        {
            long size = sender.Send(buffer);
            LogOnOffManager.RefreshLogOnOff();
            PrivateMetric("logging_client_send_data_size", size);
        }

        public void Debug(string message)
        {
            Debug(string.Empty, message);
        }

        public void Debug(string title, string message)
        {
            Debug(title, message, null);
        }

        public void Debug(string title, string message, Dictionary<string, string> tags)
        {
            Log(title, message, tags, LogLevel.Debug);
        }

        public void Info(string message)
        {
            Info(string.Empty, message, null);
        }

        public void Info(string title, string message)
        {
            Info(title, message, null);
        }

        public void Info(string title, string message, Dictionary<string, string> tags)
        {
            Log(title, message, tags, LogLevel.Info);
        }

        public void Warn(string message)
        {
            Warn(string.Empty, message);
        }

        public void Warn(string title, string message)
        {
            Warn(title, message, null);
        }

        public void Warn(string title, string message, Dictionary<string, string> tags)
        {
            Log(title, message, tags, LogLevel.Warm);
        }

        public void Error(string message)
        {
            Error(string.Empty, message, null);
        }

        public void Error(string title, string message)
        {
            Error(title, message, null);
        }

        public void Error(string title, string message, Dictionary<string, string> tags)
        {
            Log(title, message, tags, LogLevel.Error);
        }

        public void Error(Exception ex)
        {
            this.Error(ex, null);
        }

        public void Error(string title, Exception ex)
        {
            this.Error(title, ex, null);
        }

        public void Error(Exception ex, Dictionary<string, string> tags)
        {
            this.Error(ex.Message, ex, tags);
        }

        public void Error(string title, Exception ex, Dictionary<string, string> tags)
        {
            string err_msg = string.Empty;
            if (HttpContext.Current == null)
            {
                err_msg = new Error(ex, null).ToString();
            }
            else
            {
                try
                {
                    err_msg = new Error(ex, new HttpContextWrapper(HttpContext.Current)).ToString();
                }
                catch
                {
                    err_msg = new Error(ex, null).ToString();
                }
            }
            this.Error(title, err_msg, tags);
        }



        private string GetExceptionMessage(Exception ex)
        {
            StringBuilder sb = new StringBuilder();
            sb.AppendLine(ex.Message);
            sb.AppendLine(ex.Source);
            sb.AppendLine(ex.StackTrace);
            //if (ex.TargetSite != null)
            //{
            //    sb.AppendLine(ex.TargetSite.Name);
            //}
            if (ex.InnerException != null)
            {
                string msg = GetExceptionMessage(ex.InnerException);
                sb.AppendLine(msg);
            }
            return sb.ToString();
        }

        private static void PrivateMetric(string name, double value, Dictionary<string, string> tags = null, DateTime time = default(DateTime))
        {
            if (!LoggingEnabled) { return; }
            var Metric = new MetricEntity();
            Metric.Name = name;
            Metric.Value = value;
            Metric.Tags = tags;
            if (time == default(DateTime))
            {
                Metric.Time = Utils.GetUnixTime(DateTime.Now);
            }
            else
            {
                Metric.Time = Utils.GetUnixTime(time);
            }
            block.Enqueue(Metric);
        }

        public void Metric(string name, double value, Dictionary<string, string> tags = null)
        {
            PrivateMetric(name, value, tags);
        }



        public void Metric(string name, Dictionary<string, string> tags = null)
        {
            Metric(name, 1, tags);
        }

        protected LogEntity CreateLog(string source, string title, string message, Dictionary<string, string> tags, LogLevel level)
        {
            LogEntity log = new LogEntity();
            log.Level = level;
            log.Message = message;
            log.Tags = tags;
            log.Time = Utils.GetTimeTicks(DateTime.Now);
            log.Title = title;
            log.Source = source;
            log.Thread = Thread.CurrentThread.ManagedThreadId;
            if (log.Tags == null)
            {
                log.Tags = new Dictionary<string, string>();
            }
            log.Tags.Add("traceid", Trace.CorrelationManager.ActivityId.ToString("N"));
            return log;
        }

        protected virtual void Log(string title, string message, Dictionary<string, string> tags, LogLevel level)
        {
            if (!LoggingEnabled) { return; }

            LogOnOff onOff = LogOnOffManager.GetLogOnOff();

            if (level == LogLevel.Debug && onOff.Debug != 1) { return; }
            if (level == LogLevel.Info && onOff.Info != 1) { return; }
            if (level == LogLevel.Warm && onOff.Warm != 1) { return; }
            if (level == LogLevel.Error && onOff.Error != 1) { return; }

            LogEntity log = this.CreateLog(Source, title, message, tags, level);
            block.Enqueue(log);
        }

        public string GetLogs(long start, long end, int appId, int[] level = null, string title = "", string msg = "", string source = "", string ip = "", Dictionary<string, string> tags = null, int limit = 100)
        {
            string url = Settings.LoggingServerHost + "/LogViewer.ashx";

            StringBuilder query = new StringBuilder(url);
            query.Append("?start =" + start);
            query.Append("&");
            query.Append("end=" + end);
            query.Append("&");
            query.Append("appId=" + appId);
            if (level != null && level.Length > 0)
            {
                query.Append("&");
                query.Append("level=" + string.Join(",", level));
            }
            query.Append("&");
            query.Append("title=" + title);
            query.Append("&");
            query.Append("msg=" + msg);
            query.Append("&");
            query.Append("source=" + source);
            query.Append("&");
            query.Append("ip=" + ip);

            if (tags != null && tags.Count > 0)
            {
                string tags_str = string.Empty;
                foreach (var item in tags)
                {
                    tags_str += item.Key + "=" + item.Value;
                    tags_str += ",";
                }
                tags_str = tags_str.TrimEnd(',');
                query.Append("&");
                query.Append("tag=" + tags_str);
            }

            query.Append("&");
            query.Append("limit=" + limit);

            string resp = string.Empty;
            using (WebClient _client = new WebClient())
            {
                byte[] resp_byte = _client.DownloadData(query.ToString());
                resp = Encoding.UTF8.GetString(resp_byte);
            }
            return resp;
        }

        public void DebugWithTags(string title, string message, string[] tags)
        {
            var tags_dic = new Dictionary<string, string>();

            foreach (var tag in tags)
            {
                string tag_key = tag.Split('=')[0];
                string tag_val = tag.Substring(tag_key.Length, tag.Length - tag_key.Length);
                tags_dic[tag_key] = tag_val;
            }

            this.Debug(title, message, tags_dic);
        }

        public void InfoWithTags(string title, string message, string[] tags)
        {
            var tags_dic = new Dictionary<string, string>();

            foreach (var tag in tags)
            {
                string tag_key = tag.Split('=')[0];
                string tag_val = tag.Substring(tag_key.Length, tag.Length - tag_key.Length);
                tags_dic[tag_key] = tag_val;
            }

            this.Info(title, message, tags_dic);
        }

        public void WarnWithTags(string title, string message, string[] tags)
        {
            var tags_dic = new Dictionary<string, string>();

            foreach (var tag in tags)
            {
                string tag_key = tag.Split('=')[0];
                string tag_val = tag.Substring(tag_key.Length, tag.Length - tag_key.Length);
                tags_dic[tag_key] = tag_val;
            }

            this.Warn(title, message, tags_dic);
        }

        public void ErrorWithTags(string title, string message, string[] tags)
        {
            var tags_dic = new Dictionary<string, string>();

            foreach (var tag in tags)
            {
                string tag_key = tag.Split('=')[0];
                string tag_val = tag.Substring(tag_key.Length, tag.Length - tag_key.Length);
                tags_dic[tag_key] = tag_val;
            }

            this.Error(title, message, tags_dic);
        }

        public void Metric(string name, double value, DateTime time, Dictionary<string, string> tags = null)
        {
            PrivateMetric(name, value, tags, time);
        }

        public void Flush()
        {
            block.Flush();
        }
    }
}